spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaClientAutoConfiguration.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.spring;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.client.WebClientBuilder;
+
+/**
+ * An auto-configuration for {@link WebClient}.
+ */
+@Configuration
+@ConditionalOnClass(WebClient.class)
+@ConditionalOnMissingBean(WebClientBuilder.class)
+public class ArmeriaClientAutoConfiguration extends AbstractArmeriaClientAutoConfiguration {
+
+    /**
+     * Create an {@link WebClientBuilder} bean.
+     */
+    @Bean
+    public WebClientBuilder webClientBuilder(Optional<List<ArmeriaClientConfigurator>> configurators) {
```

Comment:
Hmm - how does this bean work? Since a builder is mutable it seems weird to have it exposed as a bean for any code to mutate and cause undefined behavior. Should the bean be `Supplier<WebClientBuilder>`? I guess there could also be a `WebClient` without any base URL

Reply:
As mentioned, it provides users with incomplete beans, which can lead to undefined behavior. But I thought it was okay because there is  [a bean](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/client/RestTemplateAutoConfiguration.java#L72) that Spring Boot also provides similarly and can be inferred.

Reply:
IIRC, `ArmeriaClientConfigurator` was for configuring the Armeria `WebClient` which is used internally to send the request using Spring `WebClient`.
But we don't use Armeria `WebClient` when we send a request using `RestTemplate`.
So probably we don't need this? A user can configure Armeria `WebClient` if he/she wants to use the client. 🤔

Reply:
For reference, I use this pattern to inject factories with dependencies like MeterRegistry.

https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/armeria/ClientBuilderFactory.java#L69

I think it's pretty useful. But in hindsight `Supplier<ClientBuilder>` is probably basically the same.

Reply:
Or we can change Spring bean scope to prevent undefined behavior. The default scope is a singleton, but we change it to a prototype.